### PR TITLE
[46] [Integrate] As a user, I can logout

### PR DESCRIPTION
**Pull request by manh-t**
https://github.com/manh-t/flutter-survey/issues/46

## What happened 👀

- Clear all local tokens and related properties
- Push user back to Login screen.

## Proof Of Work 📹

https://user-images.githubusercontent.com/60863885/120892617-739c7100-c639-11eb-9e30-e9d1858a0787.mp4

### CODE REVIEW

File: lib/pages/dashboard/dashboard_page.dart [hunk 1]
```diff
@@ -20,16 +25,24 @@ class DashboardPage extends StatelessWidget {
...
-            child: Consumer(
-              builder: (context, ScopedReader watch, _) {
-                final dashboardState = watch(dashboardModelProvider);
-                return Dashboard(dashboardState.surveys);
-              },
+      body: ProviderListener<DashboardState>(
+        provider: dashboardModelProvider,
+        onChange: (context, state) {
+          if (state.logout == true) {
+            Navigator.of(context).pushReplacementNamed(Routes.STARTUP_PAGE);
```
Comment (markgravity) [inline lib/pages/dashboard/dashboard_page.dart:32]: Can [it](https://github.com/manh-t/flutter-survey/blob/506672d3e89693f841e120e463b94e7d4c9d9f00/lib/pages/startup/startup_page.dart#L11) lead to some problem?

As far as I understand, the code below will watch `sharedPreferenceHelper.isLoggedIn` changes and rebuild, right?
So if it's correct, when you logout, the `sharedPreferenceHelper.isLoggedIn` will be updated and your `StartupPage`  updates with `LoginPage` 🤔 
```dart
  Widget build(BuildContext context, ScopedReader watch) {
    final sharedPreferenceHelper = watch(sharedPreferencesProvider);
    if (sharedPreferenceHelper.isLoggedIn) {
      return DashboardPage();
    }
    return LoginPage();
  }
```

then we may don't need this line
```dart
Navigator.of(context).pushReplacementNamed(Routes.STARTUP_PAGE);
```

Comment (manh-t) [reply lib/pages/dashboard/dashboard_page.dart]: @markgravity For the above code, we only watch `sharedPreferenceHelper` instance, not the state of `isLoggedIn` variable, so `isLoggedIn` variable updated won't rebuild the widget tree above 🙏

Comment (manh-t) [reply lib/pages/dashboard/dashboard_page.dart]: sorry I missed your comment in somewhere 😓

File: lib/pages/dashboard/dashboard_model.dart [hunk 2]
```diff
@@ -82,6 +85,18 @@ class DashboardModel extends StateNotifier<DashboardState> {
...
   }
 
+  void logout() {
+    state = DashboardState.loading();
+    _logoutUseCase.call().then((result) {
+      if (result is Success) {
+        state = DashboardState.clear();
+      } else {
+        state = DashboardState.error(NetworkExceptions.getErrorMessage(
+            (result as Failed).exception.networkExceptions));
```
Comment (sleepylee) [inline lib/pages/dashboard/dashboard_model.dart:95]: I think even if it's failed, we should clear all the state as well 🤔  (failed positive case). just show a message to inform users about this failure is fine enough imo.

Comment (manh-t) [reply lib/pages/dashboard/dashboard_model.dart:95]: @sleepylee did you mean we still kick the user out to the Login page although there is an error, and show a message to let them know the error? 🙏

Comment (sleepylee) [reply lib/pages/dashboard/dashboard_model.dart:95]: yep @manh-t

Comment (manh-t) [reply lib/pages/dashboard/dashboard_model.dart:95]: agree, I'll update it, thank you 😄

**Comment by manh-t** [comment]
@sleepylee @markgravity Updated 🙏

File: lib/pages/dashboard/dashboard_model.dart [hunk 3]
```diff
@@ -82,6 +85,17 @@ class DashboardModel extends StateNotifier<DashboardState> {
...
   }
 
+  void logout() {
+    state = DashboardState.loading();
+    _logoutUseCase.call().then((result) {
+      final String? error = result is Success
+          ? null
+          : NetworkExceptions.getErrorMessage(
+              (result as Failed).exception.networkExceptions);
+      state = DashboardState.clear(error);
```
Comment (sleepylee) [inline lib/pages/dashboard/dashboard_model.dart:95]: 👌

**Review by sleepylee — APPROVED** [review]

**Review by markgravity — APPROVED** [review]
LGTM 🤖

**Pull request merged**